presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -112,4 +123,51 @@ public static boolean isEqualComparisonExpression(Expression expression)
     {
         return expression instanceof ComparisonExpression && ((ComparisonExpression) expression).getOperator() == ComparisonExpression.Operator.EQUAL;
     }
+
+    static Optional<EnumType> tryResolveEnumLiteralType(QualifiedName qualifiedName, TypeManager typeManager)
+    {
+        int nParts = qualifiedName.getParts().size();
+        if (nParts < 2) {
+            // an enum literal should be of the form `MyEnum.my_key`
+            return Optional.empty();
+        }
+        try {
+            Type baseType = typeManager.getType(parseTypeSignature(qualifiedName.getParts().get(nParts - 2)));
+            if (baseType instanceof EnumType) {
+                return Optional.of((EnumType) baseType);
+            }
+        }
+        catch (IllegalArgumentException e) {
+            return Optional.empty();
+        }
+        return Optional.empty();
+    }
+
+    public static boolean isEnumLiteral(DereferenceExpression node, Type nodeType)
+    {
+        if (!(nodeType instanceof EnumType)) {
+            return false;
+        }
+        QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
+        if (qualifiedName == null) {
+            return false;
+        }
+        int nParts = qualifiedName.getParts().size();
+        return nParts >= 2
+                && qualifiedName.getParts().get(nParts - 2).equalsIgnoreCase(nodeType.toString());
+    }
+
+    public static Object resolveEnumLiteral(DereferenceExpression node, EnumType enumType)
+    {
+        QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
+        if (qualifiedName == null || !isEnumLiteral(node, enumType)) {
+            throw new SemanticException(INVALID_LITERAL, node, "Node is not an enum literal");
+        }
+        String enumKey = qualifiedName.getSuffix().toLowerCase(ENGLISH);
+        Object enumValue = enumType.getEnumMap().get(enumKey);
+        if (enumValue == null) {
+            throw new SemanticException(INVALID_LITERAL, node, format("No key '%s' in enum '%s'", enumKey, enumType.getDisplayName()));
```

Comment:
This function is too low level to throw an `SemanticException`. If you expect this function to be only called when node is a valid enum, it should throw `IllegalArgumentException`. You can change these logic to a `checkArgument`.

Reply:
Is it OK to have errors triggered by a `checkArgument` surfaced to users?
This is the only place where we check whether the provided key exists in the enum, so it will be very possible that users see this error.

Reply:
If it's not OK, we should add a `catch` at the correct level and translate the `IllegalArgumentException` to a `SemanticException`.